### ISSUE TITLE
I had an out of memory processing a huge case

### ISSUE THREAD

**Issue by hauck-jvsh** [sepinf-inc/IPED#1646]
This out of memory occurs when parsing the msgstore databases. There are several huge msgstore, with approximately 2 GB each. The problem did not occur when I disable the <param name="recoverDeletedRecords" type="bool">false</param.
I have a thread dump, but I can't figure out what was the problem.

**Comment by lfcnassif** [comment]
Isn't this a duplicate of #1364?

**Comment by lfcnassif** [comment]
Could you check if the database schema is the old one (with the messageS table, not the message one)?

**Comment by hauck-jvsh** [comment]
I check and the database is using the old schema, but the some of the dbs are corrupted.

**Comment by lfcnassif** [comment]
Maybe it is caused by recover records feature, maybe by mergeDBs, maybe both together or other together. Just the heap dump would tell us...

**Comment by hauck-jvsh** [comment]
It finishes with successful with the mergeDB enable and with the recover records disable. I can try with the recover enable and merge disable, as maybe a problem with merging corrupted messages.

**Comment by hauck-jvsh** [comment]
What do you use to analise the hprof, I tried with java visual VM and it is computing retained sizes since yesterday.

**Comment by lfcnassif** [comment]
Eclipse memory analyzer plugin, much better, you just need to increase the Xmx eclipse value.

**Comment by lfcnassif** [comment]
Hi @hauck-jvsh, could you confirm if this is a duplicate of #1364 or not?

**Comment by hauck-jvsh** [comment]
I cannot confirm because in this case the db is the old one. After digging into the problem I found that the problem is probably a combination of both, the merge and the undelete, as the undelete recovers a lot of items, and all itens found in several dbs are being held memory for the merge. This may be solved by adding processing time, instead of processing all dbs in parallel, I can change it to process only the master DB and each backup be processed and merged sequentially. This will reduce the memory requirements, but will process all db sequentially. What do you think?

**Comment by lfcnassif** [comment]
Is it possible to do a mixed approach? If the estimated memory to be used is less than x% of max heap memory, process in parallel, if it is greater than x% process them sequentially?

Since we already have an open issue to optimize memory usage by whatsapp records undelete, I'm changing the title of this to focus on the merge DBs feature memory usage here.

**Comment by lfcnassif** [comment]
I think it is a reasonable approach.
